src/arch/arm/object/smc.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2021, DornerWorks Ltd.
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ */
+#include <config.h>
+
+#ifdef CONFIG_ALLOW_SMC_CALLS
+#include <arch/object/smc.h>
+
+compile_assert(n_msgRegisters_less_than_smc_regs, n_msgRegisters <= NUM_SMC_REGS);
+
+static exception_t invokeSMCCall(word_t *buffer, bool_t call)
+{
+    word_t i;
+    seL4_Word arg[NUM_SMC_REGS];
+
+    for (i = 0; i < NUM_SMC_REGS; i++) {
+        arg[i] = getSyscallArg(i, buffer);
+    }
+
+    register seL4_Word r0 asm("x0") = arg[0];
+    register seL4_Word r1 asm("x1") = arg[1];
+    register seL4_Word r2 asm("x2") = arg[2];
+    register seL4_Word r3 asm("x3") = arg[3];
+    register seL4_Word r4 asm("x4") = arg[4];
+    register seL4_Word r5 asm("x5") = arg[5];
+    register seL4_Word r6 asm("x6") = arg[6];
+    register seL4_Word r7 asm("x7") = arg[7];
+    asm volatile("smc #0\n"
+                 : "+r"(r0), "+r"(r1), "+r"(r2), "+r"(r3)
+                 : "r"(r4), "r"(r5), "r"(r6), "r"(r7)
```

Comment:
According to DEN0028E_SMC_Calling_Convention_1.4eac0:

- A Function Identifier is passed in register W0.
- Arguments are passed in registers W1-W7.
- Results are returned in W0-W7.
- Registers W4-W7 must be preserved unless they contain results, as specified in the function definition.

(Same for 64-bit calls, but with X0-X17 instead.)

So I think both the output and input lists should be extended to the r0-r7 registers, both with `+r`.

Also, if only the 32-bit calling convention is supported, the code should check that bit 30 is not set in `smc_func_id`.

Reply:
So we are actually targeting a specific version of the SMCCC spec. v1.0. This document can be viewed here: https://developer.arm.com/documentation/den0028/a/?lang=en (The `a` in the URL should give you the correct version when you download.). You can query the SMCCC version number on your platform with an SMC call. We did our development on a Zynq UltraScale+ MPSoC, which implements spec v1.0. 

In this version, r4-r7 is just scratch space, so it won't ever be a result. 
Also, since we are using `x0-x7`, we are actually using the 64-bit calling convention.

This does mean that if future processors conform to newer calling convention versions, than we need to create version  specific implementations.

Reply:
We don't want version specific implementations. This is a new multi-platform feature, it should be as future proof as possible, which means following the latest spec. If in the future a platform does return data in registers x4-x7, with the current code it will silently clobber a register and that will be extremely hard to debug if it causes weird problems.

If you do follow the 64-bit calling convention, a lot more registers get clobbered, see table 3-1. The normal procedure call conventions are used, registers `x0-x17`  may be modified. On the upside, you don't have to check bit 30, as you can do 32-bit calls with the 64-bit convention.

With that in mind, maybe it's better to add a `do_smc` function to assembly and call it from here, I think that will get all clobbers and assumptions right automatically.

Reply:
I'm fine with limiting it to `x0-x7` for now, as it will take ages before more is practically needed, and when the limit is reached, it's easy to extend. On the other hand, it's probably not much more work to extend it now.

>  I am not entirely sure what an assembly function gets us over the inline assembly we are already doing. The context still needs to be transferred in/out of the registers?

Doing the SMC call would be just a normal function call then. But I forgot that there is a need to transfer the result too, so the advantage would be much less, if any.

Reply:
Playing around with with using all 18 registers revealed to me that inline assembly calls can only have 30 operands, so 18 inputs + 18 outputs doesn't work. Path of least resistance is probably to just modify `r4-r7` to be `r+`, so it is an input and output. 

Or if we really wanted to support all 18 registers, we would need to write a proper assembly function. Thoughts?

Reply:
I implemented the simpler approach with my latest push.

Reply:
Almost there, you also need to add all other modified registers to the clobber list, see Table 3-1.

Reply:
I appreciate the input. This implementation is much more future-proof now.